feat(revenue-analytics): implement cache target age logic for revenue syncing

rafaeelaudibert opened the pull request:
## Problem

We're hitting a major problem with Revenue Analytics that's hurting our onboarding process. Users add their first revenue source to PostHog and then open the Revenue Analytics dashboard. We'll then query their partial data and display the results, but we also display a big banner saying that we're still syncing their data so it's still in progress. This is all fine.

What happens, however, is that people come back to the UI after 30 minutes (enough time to sync medium-sized accounts) and the banner is gone, but the charts are still displaying stale data because we've cached all those insights. They can just click the `refresh` button, but that's not obvious.

## Changes

To reduce that problem, let's be much smarter with our caching mechanism:
- when we detect we're syncing data from a revenue source, let's implement a minimal cache (just 1min to avoid flooding)
- when not syncing, rather than using the default caching period, let's look at the sync frequency for the sources and use half the shortest sync period as our cache

This will guarantee that customers should see non-stale data very frequently. We're, of course, sacrificing speed/CPU for this, but I believe it's acceptable given the increase in user satisfaction.

If we ever think this is hurting our CPU, we can find smarter ways to bust the query - for example, clear the cache when DWH syncing finishes via some kind of hook (that doesnt exist yet).

Closes #34650

rafaeelaudibert committed:
feat(revenue-analytics): implement cache target age logic for revenue data syncing

We're hitting a major problem with Revenue Analytics that's hurting our onboarding process. Users add their first revenue source to PostHog and then open the Revenue Analytics dashboard. We'll then query their partial data and display the results, but we also display a big banner saying that we're still syncing their data so it's still in progress. This is all fine.

What happens, however, is that people come back to the UI after 30 minutes (enough time to sync medium-sized accounts) and the banner is gone, but the charts are still displaying stale data because we've cached all those insights. They can just click the `refresh` button, but that's not obvious.

To reduce that problem, let's be much smarter with our caching mechanism:
- when we detect we're syncing data from a revenue source, let's implement a minimal cache (just 1min to avoid flooding)
- when not syncing, rather than using the default caching period, let's look at the sync frequency for the sources and use half the shortest sync period as our cache

This will guarantee that customers should see non-stale data very frequently. We're, of course, sacrificing speed/CPU for this, but I believe it's acceptable given the increase in user satisfaction.

If we ever think this is hurting our CPU, we can find smarter ways to bust the query - for example, clear the cache when DWH syncing finishes via some kind of hook (that doesnt exist yet).

Closes #34650

arthurdedeus commented on products/revenue_analytics/backend/hogql_queries/revenue_analytics_query_runner.py:
```diff
@@ -475,3 +476,36 @@ def _subquery_for_view(self, view: type[RevenueAnalyticsBaseView]) -> ast.Select
         else:
             raise ValueError(f"Invalid view: {view}")
+
+    # If we can detect we're syncing Revenue data for the first time, cache for just 1 minute
+    # Otherwise, cache it for half the frequency we sync data from Stripe
+    def cache_target_age(self, last_refresh: Optional[datetime], lazy: bool = False) -> Optional[datetime]:
```
nit: maybe move this to the docstring of the method?

arthurdedeus commented on products/revenue_analytics/backend/hogql_queries/revenue_analytics_query_runner.py:
```diff
@@ -475,3 +476,36 @@ def _subquery_for_view(self, view: type[RevenueAnalyticsBaseView]) -> ast.Select
+
+        # If we can detect we're syncing Revenue data for the first time, cache for just 1 minute
+        # this guarantees we'll "always" have fresh data for the first sync
+        if any(
+            schema.status == ExternalDataSchema.Status.RUNNING and schema.last_synced_at is None for schema in schemas
+        ):
```
greptile has a point here. using `any` will evaluate every expression in the iterator, while something like `next` will stop at the first occurrence that matches.

idk if this is concerning here, though. feel free to ignore.

rafaeelaudibert replied:
Well, not exactly true. It will stop iterating once it finds the right schema:

<img width="384" height="175" alt="image" src="https://github.com/user-attachments/assets/9f11cab1-d6d8-4c58-8bc0-8a601b72b7ee" />

More often than not we'll need to iterate across all events because we need the smaller sync_frequency and that requires the whole list. I'm not afraid of spending some extra time on the edge case, so I'mma keep this like this!

arthurdedeus commented on products/revenue_analytics/backend/hogql_queries/revenue_analytics_query_runner.py:
```diff
@@ -475,3 +476,36 @@ def _subquery_for_view(self, view: type[RevenueAnalyticsBaseView]) -> ast.Select
+        # In the rare case where we can't figure out the interval, default to caching for 6 hours
+        intervals = [schema.sync_frequency_interval for schema in schemas if schema.sync_frequency_interval is not None]
+        if not intervals:
+            return last_refresh + timedelta(hours=6)
```
nit: extract magic number to constant

arthurdedeus reviewed (approved):
elegant solution here, very nice

rafaeelaudibert committed:
refactor(revenue-analytics): remove magic numbers

Updated the cache target age logic in `RevenueAnalyticsQueryRunner` to use class constants for better readability and maintainability. Adjusted test cases to reference these constants instead of hardcoded values, ensuring consistency across the codebase. This change clarifies the caching strategy for first-time syncs and default scenarios.
